pkg/api/api.go

```diff
@@ -0,0 +1,410 @@
+package api
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+
+	"github.com/openshift/sippy/pkg/html"
+	"github.com/openshift/sippy/pkg/util"
+	"k8s.io/klog"
+)
+
+// PassRate describes statistics on a pass rate
+type PassRate struct {
+	Percentage          float64 `json:"percentage"`
+	ProjectedPercentage float64 `json:"projectedPercentage,omitempty"`
+	Runs                int     `json:"runs"`
+}
+
+// SummaryAcrossAllJobs describes the category failuregroups
+// valid keys are latestXDays, and prevXDays
+type SummaryAcrossAllJobs struct {
+	TestExecutions     map[string]int     `json:"testExecutions"`
+	TestPassPercentage map[string]float64 `json:"testPassPercentage"`
+}
+
+// FailureGroups describes the category failuregroups
+// valid keys are latestXDays, and prevXDays
+type FailureGroups struct {
+	JobRunsWithFailureGroup map[string]int `json:"jobRunsWithFailureGroup"`
+	AvgFailureGroupSize     map[string]int `json:"avgFailureGroupSize"`
+	MedianFailureGroupSize  map[string]int `json:"medianFailureGroupSize"`
+}
+
+// CanaryTestFailInstance describes one single instance of a canary test failure
+// passRate should have percentage (float64) and number of runs (int)
+type CanaryTestFailInstance struct {
+	Name     string   `json:"name"`
+	Url      string   `json:"url"`
+	PassRate PassRate `json:"passRate"`
+}
+
+// PassRatesByJobName is responsible for the section job pass rates by job name
+type PassRatesByJobName struct {
+	Name         string              `json:"name"`
+	Url          string              `json:"url"`
+	PassRates    map[string]PassRate `json:"passRates"`
+	FailingTests []FailingTest       `json:"failingTests"`
+}
+
+// FailingTest describes one single instance of a failed test and associated bugs
+// passRate may include projected pass rate (float64), percentage (float64), and number of runs (int)
+type FailingTest struct {
+	Name     string     `json:"name"`
+	Url      string     `json:"url"`
+	PassRate PassRate   `json:"passRate"`
+	Bugs     []util.Bug `json:"bugs"`
+}
+
+// FailingTestBug describes a single instance of failed test with bug or failed test without bug
+// differs from failingtest in that it includes pass rates for previous days and latest days
```

Comment:
can't you just use FailingTestBug everywhere since it's a superset of FailingTest?

Reply:
yeah I realized failingTest is actually wholly redundant since I'm not returning bug data, so I deleted it.